Updated mp_builtin_pow() to use the optimised mpz_pow3_inpl() function.

nickovs opened the pull request:
I took a stab at fixing #2774 so that the built in pow() function is usable with arguments large enough to be useful for cryptographic functions, and to not use exponential memory and time. All it does is make use of the existing (but previously commented out) `mpz_pow3_inpl()` function written by @dpgeorge.

This is my first pull request for Micropython to please feel free to send me any comments and complaints and I will try to fix them!

Commit:
esp8266/machine_pin: Implement pin ioctl protocol.

For polymorphic interfacing on C level.

Commit:
extmod/machine_signal: Implement "signal" abstraction for machine module.

A signal is like a pin, but ca also be inverted (active low). As such, it
abstracts properties of various physical devices, like LEDs, buttons,
relays, buzzers, etc. To instantiate a Signal:

pin = machine.Pin(...)
signal = machine.Signal(pin, inverted=True)

signal has the same .value() and __call__() methods as a pin.

Commit:
esp8266/modmachine: Add Signal class.

Commit:
examples/hwapi: Use Signal for inverted LED on ESP-12.

nickovs committed:
Updated mp_builtin_pow() to use the optimised mpz_pow3_inpl() function.

nickovs committed:
py: Added optimised support for 3-argument calls to builtin.pow()

Updated modbuiltin.c to add conditional support for 3-arg calls to
pow() using MICROPY_INT_POW3 config parameter. Added support in
objint_mpz.c for for optimised implementation.

nickovs commented:
It seems that the Travis CI build is currently failing with:
```
$ (cd tests && for exp in *.exp; do testbase=$(basename $exp .exp); echo -e "\nFAILURE $testbase"; diff -u $testbase.exp $testbase.out; done)

FAILURE *
diff: *.exp: No such file or directory
diff: *.out: No such file or directory
```
I didn't touch anything in the tests since the existing tests covered the new code. I see that @dpgeorge's #2795 is failing in the same way, so I'm pretty sure it's not my fault! That said, I don't know anything about the Travis setup here so I don't know what I can do to fix this.

pfalcon commented:
Finding a cause of a Travis failure is a bit of black art, as it doesn't have quick-links to potential errors (or we don't have it configured). How I usually do it is grepping its log for "error" (many false positives) or "exited with" (not selective enough, scroll back until you find the issue).

nickovs committed:
Fixed typo in MP_OBJ_TO_PTR cast

nickovs commented:
OK, I found the error. The Travis build is clean now.

dpgeorge commented:
Thanks for the contribution!

A few general comments:
- please use only spaces (no tabs) and use a "tab width" of 4
- if statements should (almost) always use curly braces, even for 1-liners
- variable declarations should be as close as possible to their first use
- it would be great if there were tests for this, even better if it had 100% coverage (can be checked via the coveralls link in the CI output above

nickovs committed:
Extended test coverage to cover all code paths.
Adjusted indentation, formatting and code style.

nickovs commented:
OK. I have adjusted the code style to have 4 spaces, no tabs and braces on single-statement blocks and I also moved the variable definitions to be as close to use as made sense.

I also added comprehensive test coverage including the exception cases; coveralls now reports 100% on modbuiltins.c. Let me know if there is anything else you would like me to change.

dpgeorge commented on py/modbuiltins.c:
```diff
@@ -375,10 +375,44 @@ STATIC mp_obj_t mp_builtin_ord(mp_obj_t o_in) {
+#else
+        default:
+	    if (!MP_OBJ_IS_INT(args[0]) || !MP_OBJ_IS_INT(args[1]) || !MP_OBJ_IS_INT(args[2])) {
+	        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "pow() with 3 arguments requires integers"));
```
A shortcut function to use is mp_raise_TypeErro("...msg...").

nickovs replied:
Done

dpgeorge commented on py/modbuiltins.c:
```diff
@@ -375,10 +375,44 @@ STATIC mp_obj_t mp_builtin_ord(mp_obj_t o_in) {
+                mp_obj_int_t *res_p = (mp_obj_int_t *) MP_OBJ_TO_PTR(result);
+
+	        mpz_t l_temp, r_temp, m_temp;
+	        mpz_t *lhs = mp_mpz_for_int(args[0], &l_temp);
```
There's something a bit wrong with indenting here...

nickovs replied:
Fixed.

dpgeorge commented on tests/basics/builtin_pow.py:
```diff
@@ -8,4 +8,35 @@
+    print(hex(pow(y, y-1, x))) # Should be a 'big value'
+    print(hex(pow(y, y-1, y))) # Should be a 'big value'
+except OverflowError:
+    print("OverflowError, expected for builds without mpz support")
```
The test suite currently doesn't handle the case when mpz is disabled.  So you can just remove the try-except block here.

(If we do ever support tests without mpz, then such tests need to be factored into separate files, so the whole file is skipped.)

nickovs replied:
Done.

dpgeorge reviewed (commented):
Can you please also add a config option, eg `MICROPY_OPT_MPZ_POW3`, to enable/disable this?  It would be disabled by default.

pfalcon commented:
> Can you please also add a config option, eg MICROPY_OPT_MPZ_POW3, to enable/disable this? It would be disabled by default.

I thought about an option to disable pow() altogether, as 2-arg variant can be replaced by `**` operator.

dpgeorge commented:
> I thought about an option to disable pow() altogether,

pow is pretty basic and used a lot... such an option would only be for really constrained ports.

nickovs commented:
> Can you please also add a config option, eg MICROPY_OPT_MPZ_POW3, to enable/disable this? It would be disabled by default.

I can certainly do that. That said, the current original implementation throws an out-of-memory exception for anything other than trivially small exponents if you have MICROPY_LONGINT_IMPL_MPZ enabled. It is safe with MICROPY_LONGINT_IMPL_NONE and MICROPY_LONGINT_IMPL_LONGLONG since they silently truncate the intermediate results. Furthermore, you elsewhere wrote:

> The test suite currently doesn't handle the case when mpz is disabled. So you can just remove the try-except block here.

Since I've added proper test coverage for the builtins.pow() the old implementation will fail the tests since we are now testing `pow(x,y,z)` with meaningful-sized arguments. This means that if the proposed configuration option is in its default state the tests will fail!

nickovs committed:
Moved throw of TypeError to use the convenience function mp_raise_TypeError
Removed try/except around MPZ pow() tests

nickovs commented:
For what it's worth, disabling the new code on the stmhal build takes the size of firmware.elf from 636124 to 636062, a saving of 62 bytes or a little less than 0.01%.

I'm happy to put in a configuration option to disable this code but given that doing this will make reasonable inputs throw MemoryError exceptions I'd be inclined to have the switch (a) disable 3-argument pow() altogether and (b) be off by default. Having 3-argument pow() and multi-precision integers both enabled without having the efficient implementation switched on doesn't seem like a good plan to me.

pfalcon commented:
> This means that if the proposed configuration option is in its default state the tests will fail!

@dpgeorge meant that its default state in py/mpconfig.h should be false. We can enable it for Unix port if you insist ;-).

> For what it's worth, disabling the new code on the stmhal build takes the size of firmware.elf from 636124 to 636062, a saving of 62 bytes or a little less than 0.01%.

ELF files contain a lot of alignment, etc. Valid comparison would be on raw .bin files (or on size(1) called on .elf).

nickovs commented:
OK. The difference in sizes for the firmware1.bin and firmware.dfu files is 392 bytes, which is just over 0.1%.

I am still of the opinion that if you are going to include a whole multi-precision integer package in the build then it should not crash with MemoryError exceptions when fed reasonable inputs. The old 3-arg pow() implementation works just fine when you compile without mpz support; the code here only enables the new version when you have mpz. I'd be fine with a config option that disables 3-arguement pow() altogether since the result is reliable. It just seems wrong to me to have a configuration for "Enable an unreliable implementation that may or may no crash depending on your inputs", which is what the old implementation does when mpz support is present.

nickovs committed:
Added MICROPY_MPZ_POW3 configuration option to conditionally remove 3-argument pow()

stinos commented:
@nickovs can you also check https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md and make your commits adhere to it? E.g. something like 'py: Update xyz' instead of 'Updated xyz' And make the body of the commit message explain *why* the change is needed. You say 'Added MICROPY_MPZ_POW3 configuration option etc' but without stating why anyone would need that. The reason is in the comments above but should really be explained as part of the commit as that's much handier for future reference. Also take the second commit for example: the body says 'Removed try/except around MPZ pow() tests'. That is clear by reading the diff so it's merely duplicate information. Instead it should say why it's somehow ok to just remove the exception handling. Also you should rebase/squash some commits so they create an easy to follow 'story'. Can till be seperate commits, but it should mean something. 'Fix typo' commits should definitely be squashed.

dpgeorge committed:
extmod/vfs: Expose lookup_path_raw as mp_vfs_lookup_path.

It can be useful for low-level lookup of paths by ports.

dpgeorge committed:
lib/oofatfs: Update oofatfs library.

From https://github.com/micropython/oofatfs, branch work-R0.12b,
commit 46fb53331e7a583c29a41d37ce4b53f2718597e5.

dpgeorge committed:
cc3200: Convert to use new VFS sub-system and new ooFatFs library.

dpgeorge committed:
extmod/vfs_fat: Remove MICROPY_READER_FATFS component.

dpgeorge committed:
extmod/vfs_fat: Remove MICROPY_FSUSERMOUNT_ADHOC config option.

dpgeorge committed:
extmod: Remove MICROPY_FSUSERMOUNT and related files.

Replaced by MICROPY_VFS and the VFS sub-system.

dpgeorge committed:
esp8266/mpconfigport.h: Remove obsolete MICROPY_FATFS_VOLUMES config.

dpgeorge committed:
unix/mpconfigport.h: Remove obsolete MICROPY_FATFS_VOLUMES config.

dpgeorge committed:
extmod/vfs_fat: Remove MICROPY_FATFS_OO config option.

Everyone should now be using the new ooFatFs library.  The old one is no
longer supported and will be removed.

dpgeorge committed:
lib/fatfs: Remove old fatfs library component, it's replaced by oofatfs.

dpgeorge committed:
extmod: Merge old fsusermount.h header into vfs.h and vfs_fat.h.

vfs.h is for generic VFS declarations, and vfs_fat.h is for VfsFat
specific things.

dpgeorge committed:
extmod/vfs_fat: Remove unused fatfs_builtin_open function.

dpgeorge committed:
extmod/vfs_fat: Remove unused function fat_vfs_listdir.

dpgeorge committed:
extmod/vfs_fat: Update to use FF_DIR instead of DIR.

dpgeorge committed:
unix/modos: Remove VfsFat from uos module, it's now in uos_vfs module.

Commit:
docs/uio: Typo fixes/lexical improvements.

Commit:
examples/hwapi: Consistently use Signal class to define LEDs.

pfalcon commented:
@nickovs : Looking in more detail, there're more issues with this patch. The biggest one is that it leaks implementation detail of a particular object implementation into the generic module, modbuiltins.c. That module doesn't know anything about mpz_t, and shouldn't know, and even wrapped in in #if's, it's not pretty. Sorry about this. The good news is that we appreciate contributions to this project, and ready to support first-time contributors by cooperating on patch implementation, by picking off on the hard parts. 

Actually, I was ready to do refactoring myself now, but decided to ask if you have time and interest to go thru the whole process yourself actually. Note that you didn't pick up the easiest task to start with, even despite it may seem so. That's because of the high configurability of MicroPython, and specifically of this part (integer object implementation). I hit the same issue with .from_bytes() implementation - even with my experience in uPy, it took more chore than I initially expected. So, I'd recommend to handover it now to a maintainer.

dpgeorge committed:
stmhal: Use LED constants from PYBv4 onwards.

dpgeorge committed:
stmhal: Set the FatFs partition number when initialising VFS object.

stmhal has MULTI_PARTITION enabled for FatFs and so these values need to be
initialised.

nickovs commented:
@pfalcon Thank you for the feedback. I did think about that when I started the code but it seemed like if it was conditional on the configuration than this was as good as any place to put it and it would save a layer of function calls. I will move the code to objint_mpz.c, since that clearly has knowledge of mpz_t.

dpgeorge committed:
stmhal: Fix stack pointer initialisation for F411 and F429 boards.

The stack pointer should start pointing 1 byte past the top of the end of
RAM.

dpgeorge committed:
docs/pyboard/tutorial/lcd160cr_skin: Fix typo, get_touched->get_touch.

pfalcon commented:
That's the right place for it, thanks. While you refactor it, also consider applying suggestions above on the commit squashing and format of commit messages (you're welcome to add individual commits to PR if it helps to track your work while it's WIP, but for the final merge, they need to be squashed; if a contributor does that, it saves few minutes to maintainers (which usually amounts to be a patch being merged sooner when it's ready)).

dpgeorge committed:
extmod/vfs_reader: Fix use of NLR by popping context.

springltd committed:
extmod/vfs_fat.c: Use explicit include path for timeutils.h.

Commit:
zephyr/Makefile.zephyr: Support and default to networked (SLIP) QEMU.

Also works for non-networked builds (like minimal).

Commit:
tests/run-tests: Skip frozenset tests if set literal syntax is not available.

Commit:
tests/basics/zip: Make skippable.

dpgeorge committed:
stmhal/README: Add paragraph about building mpy-cross.

dpgeorge committed:
stmhal: On boot, mount all available partitions of the SD card.

The first partition is mounted as "/sd" and subsequent partitions are
mounted as "/sd<part_num>".  This is backwards compatible with the previous
behaviour, which just mounted the first partition on "/sd".

At this point, only FatFs filesystems are mounted.

dpgeorge committed:
stmhal/main: Put /sd directory before /flash in sys.path.

If the SD card is mounted then its libraries (ie those that are imported)
should override any in /flash.

dpgeorge committed:
lib/oofatfs/ffconf.h: Allow to configure FS_EXFAT option.

Using MICROPY_FATFS_EXFAT.  Enabling this has licensing implications;
see https://www.microsoft.com/en-us/legal/intellectualproperty/mtl/exfat-licensing.aspx

dpgeorge committed:
stmhal/pin: Add C-level pin ioctl method.

dpgeorge committed:
stmhal/modmachine: Add machine.Signal type.

dpgeorge committed:
stmhal/main: Guard init_sdcard_fs with MICROPY_HW_HAS_SDCARD.

dpgeorge committed:
stmhal/pendsv: Fill in comments about what the stack contains.

dpgeorge committed:
tests/thread: Make thread_exc2 runable on baremetal.

dpgeorge committed:
stmhal: Initial implementation of multithreading, currently disabled.

This patch brings the _thread module to stmhal/pyboard.  There is a very
simple round-robin thread scheduler, which is disabled if there is only
one thread (for efficiency when threading is not used).

The scheduler currently switches threads at a rate of 250Hz using the
systick timer and the pend-SV interrupt.

The GIL is disabled so one must be careful to use lock objects to prevent
concurrent access of objects.

The threading is disabled by default, one can enabled it with the config
option MICROPY_PY_THREAD to test it out.

dhylands committed:
unix: Make stack be non-executable

This PR is to address issue #2812.

dpgeorge committed:
stmhal/main: Use _estack value to initialise stack extents.

springltd committed:
lib/oofatfs/ffconf.h: Add MICROPY_FATFS_NORTC option.

dpgeorge committed:
stmhal: Fix build issue when MICROPY_PY_THREAD is disabled.

dpgeorge committed:
tests/thread: Fix stack size test so tests run reliably on baremetal.

Commit:
zephyr/README: Describe many gotchas of networked builds.

Commit:
py/objstringio: Allow to specify initial capacity by passing numeric argument.

E.g. uio.BytesIO(100) will allocate buffer with 100 bytes of space.

Commit:
tests/heapalloc_bytesio: Test for BytesIO with preallocates space.

nickovs committed:
Merge branch 'fast_pow3' of github.com:/nickovs/micropython into fast_pow3

pfalcon commented:
@nickovs : FYI, we use rebase-based workflow (just like (most of maintainers of) Linux and many other projects nowadays). Just thought I'd mention, because it's actually easier for contributors too.

nickovs commented:
OK, I've moved the mpz-specifc code into objint_mpz.c and put in place a configuration switch to disable 3-argument pow() for people who need to save the space. I also (think I have) squashed the commits so that it should be a single big commit with a meaningful, style-conformant commit message. Hopefully this will tick all the boxes!

pfalcon commented:
> I also (think I have) squashed

Something went wrong:

>  Commits 55   Files changed 100

Merge commits are to blame, they're the root of all evil. Never "git pull", always "git pull --rebase" ;-)

nickovs commented:
OK. What should I do now?

pfalcon commented:
If you actually squashed your changes into 1 commit, it should be among those 55 I guess. I'd look for it, created a fresh branch from master, cherry-picked that commit and re-pushed. If squash didn't work out, there would be multiple commits to cherry-pick and then squash. As usual, I'm happy to offer assistance with this git grunge.

nickovs commented:
OK, rather than try to wrestle with git, I have created a new branch and new pull request #2835 which should be clean. I suggest we close this one and move the discussion to there.

pfalcon commented:
You have the powers, please do.
